### ISSUE TITLE
A faster compilation model

### ISSUE THREAD

**Issue by eudoxia0** [austral/austral#213]
[quality status: stream of consciousness notes]

Efficiently compiling large programs is difficult. Compilers are constrained from various directions: we want compile times to be short, we want compilers to be massively parallel, we want final programs to be small and fast. Some of these goals are in contradiction. The "compilation model" of a programming language is how it compiles and ties into external modules.

Sadly this is an area where the compiler literature is very lacking and we have to rely on the documentation from the internals of modern compilers.

# Whole Program Compilation

The simplest compiler works like this: it takes the import DAG of the program, does a [topological sort](https://en.wikipedia.org/wiki/Topological_sorting) to get a linear order, and loads and compiles each module, one at a time, from the most distant leaf dependency to the program entrypoint.

For example, consider the following import DAG, where each box is a module:

![Dependency DAG](https://user-images.githubusercontent.com/1612511/173173759-852ceed4-1386-4944-bb41-bae8244c8e0e.png)

Graphviz `dot` source:

```dot
digraph deps {
    app->sql;
    app->http;
    sql->socket;
    http->socket;
    sql->iter;
    sql->parser;
    iter->stdlib;
    parser->stdlib;
    socket->stdlib;
}
```

A topological sort would be:

```
1. stdlib
2. parser
3. iter
4. socket
5. sql
6. http
7. app
```

Each module can import things from a module defined earlier.

Austral, as it works at present, is a whole-program compiler, because the bootstrapping compiler was designed for simplicity. However this has many drawbacks:

1. Memory usage: every single declaration in the dependency DAG has to be stored in memory.
2. Time: every single change that is made involves compiling the entire universe (including the standard library, when it exists) from scratch.

When the compiler is done, it spits out a giant C file, and the process has to be repeated again. Naturally, we want a better way to do this.

# Resumable Cutoff Compilation

This is an intermediate step, in terms of complexity of implementation and performance, between whole program compilation and separate compilation.

Looking at the compilation order above:

```
1. stdlib
2. parser
3. iter
4. socket
5. sql
6. http
7. app
```

The idea is to split it into chunks: on the first chunk, we compile the chunk's modules one after the other, and serialize the environment, in the second chunk, we load the serialized environment, and compile that chunk's modules, then dump the new, larger environment.

The tradeoff here is that each chunk gets slightly slower to compile since the serialized environment you have to load grows monotonically. You can reduce this by trimming private types/functions from the environment just before serializing. Also, the performance penalty might be worth it, since you can serialize the table of monomorphs across compilations, and reduce the need to compile the same monomorph multiple times.

The result of compiling each chunk is the serialized env, a header file and a C source code file. The header file starts with an `#include` directive to the header file from the previous chunk. When a chunk is compiled, we compile the C source code file to an object file. We still have to pay the cost of recursive includes.

```
chunk 1:
    - compile stdlib
    - dump env-1.json, chunk-1.h, chunk-1.c
    - gcc -c chunk-1.c -o chunk-1.o

chunk 2:
    - load env-1.json
    - compile parser
    - compile iter
    - compile socket
    - dump env-2.json, chunk-2.h, chunk-2.c
    - gcc -c chunk-2.c -o chunk-2.o

chunk 3:
    - load env-2.json
    - compile sql
    - compile http
    - dump env-3.json, chunk-3.h, chunk-3.c
    - gcc -c chunk-3.c -o chunk-3.o

chunk 4:
    - load env-3.json
    - compile app
    - dump -c chunk-4.h, chunk-4.c
    - gcc -c chunk-4.c -o chunk-4.o
    - gcc chunk-1.o chunk-2.o chunk-3.o chunk-4.o -o app 
```

# Separate Compilation

The goal is fully separate compilation. In separate compilation, when the compiler compiles a module, it spits out both object code and a kind of serialized (possibly binary, the goal is it should be fast to load) form of the module, containing everything another invocation of the compiler needs to use that module. Call it a _module metadata_ file, it doesn't matter whether it's JSON or binary or whatever, that's just an implementation detail. A module metadata file contains:

- The name of the module.
- The names of public functions (and, in the case of generic functions, the bodies, since generic functions may be imported and instantiated)
- The definition of public types.
- The type signatures of opaque types (and their sizes in bytes, probably).

And so on.

There are a number of difficulties with separate compilation, mainly with transitive dependencies: a module `A` may contain a type that is an alias of a type imported from `B`, and said module may contain a type alias to a type imported from `C`, and so on.

So the module metadata file of a module is not just the public contents of the module. It also has to refer to other modules, and there are two ways of doing this:

1. **Links:** the module metadata file links  to other module metadata files, that must also be parsed (which themselves contain links, and so on) in which case we have the same problem we have with C header files: the header files have to be recursively loaded from the file system, which can be slow if the filesystem is slow.

2. **Recursive embedding:** when we dump a module `Foo` into its module metadata file, we recursively embed every generic function body, generic type definition, etc. that might potentially be accessible from `Foo`. Here we're trading multiple recursive file accesses for giant, slow-loading module metadata file.

It's debatable what is better.

A "sufficiently smart" compiler will trim things that are definitely not accessible when dumping module metadata.

It might also be desirable to make the module metadata files lazy-loading: so if module `A` has hundreds of transitive dependencies, not every invocation of the compiler has to load them. Rather, the compiler will only load module metadata for transitive dependencies on an as-needed basis, and cache along the way to make future accesses faster.

With fully separate compilation we can take advantage of parallelism: every "horizontal" level of the dependency DAG can be built in parallel. In the example above:

1. `stdlib` is built first.
2. When `stdlib` is done building, then `socket`, `iter` and `parser` are built in parallel.
3. When `socket` is done, `http` starts building.
4. When `socket`, `iter`, and `parser` are done building, then `sql` starts building.
5. When `http` and `sql` are done building, then `app` is built.

In the large this improves performance.

# Prior Art

- [Rust](https://www.rust-lang.org/) has a concept of "crate metadata", which is the serialized contents of a crate. See: https://github.com/rust-lang/rust/blob/75307c22f33cb47966466ba1daca08573a3c583e/compiler/rustc_metadata/src/locator.rs . Note especially:

```
## Transitive Dependencies

Let's say we've got three crates: A, B, and C. A depends on B, and B depends
on C. When we're compiling A, we primarily need to find and locate B, but we
also end up needing to find and locate C as well.

The reason for this is that any of B's types could be composed of C's types,
any function in B could return a type from C, etc. To be able to guarantee
that we can always type-check/translate any function, we have to have
complete knowledge of the whole ecosystem, not just our immediate
dependencies.

So now as part of the "find a corresponding crate on the filesystem" step
above, this involves also finding all crates for *all upstream
dependencies*. This includes all dependencies transitively.
``` 

- [Nim](https://nim-lang.org/) uses "rodfiles": https://nim-lang.org/docs/compiler/ic/rodfiles.html

- [Ada](https://en.wikipedia.org/wiki/GNAT) (well, GNAT) uses `.ali` (Ada Library Information) files: https://gcc.gnu.org/onlinedocs/gnat_ugn/The-Ada-Library-Information-Files.html

# Tasks

- #215 
- #219
- #220 
- #221
- #222

**Comment by eudoxia0** [comment]
Closing because this is distant and vague and clutters up the issue list. The right place for this is the roadmap.
